File: admin/Admi/firebase/functions/driver_registration_document_status.js
/**
 * Registration V2 document completeness — shared with Admin/Driver Dart helper.
 * Required slots: national_id, vehicle_registration, driver_license (front+back).
 * Profile photo is checked separately for submit/approve blockers.
 */
'use strict';

const REQUIRED_TYPES = [
  'national_id',
  'vehicle_registration',
  'driver_license',
];

const TYPE_KEYS = {
  national_id: ['doc_national_id', 'img_id_rksh'],
  vehicle_registration: ['doc_vehicle_registration', 'img_id_car'],
  driver_license: ['doc_driver_license_front', 'doc_driver_license_back', 'doc_driver_license'],
};

function isStoragePath(raw) {
  const p = String(raw || '').trim();
  return p.startsWith('users/') && !p.includes('..');
}

function isHttps(url) {
  return typeof url === 'string' && /^https:\/\//i.test(url.trim());
}

function storagePathFrom(data, v2Key) {
  const slot = data[v2Key];
  if (slot && typeof slot === 'object') {
    const p = String(slot.storagePath || '').trim();
    if (isStoragePath(p)) return p;
  }
  return '';
}

function urlFrom(data, v2Key, legacyKey) {
  const slot = data[v2Key];
  if (slot && typeof slot === 'object' && typeof slot.url === 'string') {
    const u = slot.url.trim();
    if (u) return u;
  }
  if (legacyKey) {
    const leg = data[legacyKey];
    if (typeof leg === 'string' && leg.trim()) return leg.trim();
  }
  return '';
}

function slotStatusRaw(data, v2Key) {
  const slot = data[v2Key];
  if (slot && typeof slot === 'object' && typeof slot.status === 'string') {
    return String(slot.status).trim().toLowerCase();
  }
  return '';
}

function sidePresent(data, v2Key, legacyKey) {
  if (storagePathFrom(data, v2Key)) return true;
  const url = urlFrom(data, v2Key, legacyKey);
  return url && (isStoragePath(url) || isHttps(url));
}

function isApprovedLegacyLicenseOnly(data) {
  const status = String(data.registration_status || '').trim();
  const approved = status === 'approved' && data.actev_mndob === true;
  const legacy = sidePresent(data, 'doc_driver_license', '');
  const front = sidePresent(data, 'doc_driver_license_front', '');
  const back = sidePresent(data, 'doc_driver_license_back', '');
  return approved && legacy && !front && !back;
}

function driverLicenseSubmitOk(data, countryRequirements) {
  const backRequired = isLicenseBackRequired(countryRequirements);
  const front = sidePresent(data, 'doc_driver_license_front', '');
  const back = sidePresent(data, 'doc_driver_license_back', '');
  if (front && (!backRequired || back)) return true;
  return false;
}

function isLicenseBackRequired(countryRequirements) {
  if (!countryRequirements || typeof countryRequirements !== 'object') return true;
  const cfg = countryRequirements.driverLicenseBack || countryRequirements.driver_license_back;
  if (cfg && typeof cfg === 'object' && cfg.required === false) return false;
  return true;
}

function statusForType(data, type) {
  const keys = TYPE_KEYS[type];
  if (!keys) return 'missing';
  if (type === 'driver_license') {
    for (const key of keys) {
      const raw = slotStatusRaw(data, key);
      if (raw === 'rejected') return 'rejected';
      if (raw === 'needs_reupload') return 'needs_reupload';
    }
    const front = sidePresent(data, 'doc_driver_license_front', '');
    const back = sidePresent(data, 'doc_driver_license_back', '');
    if (front && back) return 'complete';
    if (isApprovedLegacyLicenseOnly(data)) return 'complete';
    const legacy = sidePresent(data, 'doc_driver_license', '');
    if (legacy && !front && !back) return 'missing';
    return 'missing';
  }
  const [v2, legacy] = keys;
  const raw = slotStatusRaw(data, v2);
  if (raw === 'rejected') return 'rejected';
  if (raw === 'needs_reupload') return 'needs_reupload';
  if (storagePathFrom(data, v2)) return 'complete';
  const url = urlFrom(data, v2, legacy);
  if (url && (isStoragePath(url) || isHttps(url))) return 'complete';
  return 'missing';
}

function profilePhotoOk(data) {
  if (isStoragePath(data.photo_storage_path)) return true;
  return isHttps(data.photo_url || '');
}

function overallRequiredDocs(data) {
  const slots = REQUIRED_TYPES.map((t) => statusForType(data, t));
  if (slots.some((s) => s === 'rejected' || s === 'needs_reupload')) {
    return 'needs_reupload';
  }
  if (slots.some((s) => s === 'missing')) return 'missing';
  return 'complete';
}

function registrationDocumentsStatus(data) {
  const flow = Number(data.registration_flow_version || 0);
  if (flow !== 2) return 'unknown_legacy';
  return overallRequiredDocs(data);
}

function isCompleteForSubmit(data) {
  return overallRequiredDocs(data) === 'complete' && profilePhotoOk(data);
}

function phonePresent(data) {
  const pn = String(data.phone_number || data.phoneNumber || '').trim();
  if (pn.replace(/\D/g, '').length >= 8) return true;
  const phoneN = data.phone_n ?? data.phoneN;
  if (typeof phoneN === 'number' && String(Math.trunc(phoneN)).length >= 8) {
    return true;
  }
  if (typeof phoneN === 'string' && phoneN.replace(/\D/g, '').length >= 8) {
    return true;
  }
  return false;
}

module.exports = {
  REQUIRED_TYPES,
  statusForType,
  profilePhotoOk,
  overallRequiredDocs,
  registrationDocumentsStatus,
  isCompleteForSubmit,
  phonePresent,
  driverLicenseSubmitOk,
  isLicenseBackRequired,
  sidePresent,
  isApprovedLegacyLicenseOnly,
};

File: admin/Admi/lib/core/driver_registration_document_status.dart
import '/core/driver_license_document.dart';

/// Canonical Registration V2 document completeness (Admin + Driver shared logic).
///
/// Required V2 slots: national_id, vehicle_registration, driver_license (front+back).
/// Profile photo is also required for submit (kept separate in [profilePhotoOk]).
///
/// V2 SoT: `doc_*.storagePath` under `users/{uid}/...`.
/// Legacy records may still use HTTPS URL strings only.
enum DriverRegistrationDocStatus {
  complete,
  missing,
  needsReupload,
  rejected,
}

class DriverRegistrationDocumentSlotStatus {
  const DriverRegistrationDocumentSlotStatus({
    required this.type,
    required this.status,
    required this.hasAsset,
    this.usesStoragePath = false,
  });

  final String type;
  final DriverRegistrationDocStatus status;
  final bool hasAsset;
  final bool usesStoragePath;
}

abstract final class DriverRegistrationDocumentStatus {
  DriverRegistrationDocumentStatus._();

  static const requiredTypes = <String>[
    'national_id',
    'vehicle_registration',
    'driver_license',
  ];

  static bool isStoragePath(String? raw) {
    final p = (raw ?? '').trim();
    return p.startsWith('users/') && !p.contains('..');
  }

  static String? _storagePathFrom(Map<String, dynamic> data, String v2Key) {
    final slot = data[v2Key];
    if (slot is Map && slot['storagePath'] is String) {
      final p = (slot['storagePath'] as String).trim();
      if (isStoragePath(p)) return p;
    }
    return null;
  }

  static String? _urlFrom(
      Map<String, dynamic> data, String v2Key, String legacyKey) {
    final slot = data[v2Key];
    if (slot is Map && slot['url'] is String) {
      final u = (slot['url'] as String).trim();
      if (u.isNotEmpty) return u;
    }
    final leg = data[legacyKey];
    if (leg is String && leg.trim().isNotEmpty) return leg.trim();
    return null;
  }

  static bool _hasPresentAsset(
    Map<String, dynamic> data,
    String v2Key,
    String legacyKey,
  ) {
    if (_storagePathFrom(data, v2Key) != null) return true;
    final url = _urlFrom(data, v2Key, legacyKey);
    return url != null && url.startsWith('https://');
  }

  static String _slotStatusRaw(Map<String, dynamic> data, String v2Key) {
    final slot = data[v2Key];
    if (slot is Map && slot['status'] is String) {
      return (slot['status'] as String).trim().toLowerCase();
    }
    return '';
  }

  static DriverRegistrationDocStatus statusForType(
    Map<String, dynamic> data,
    String type,
  ) {
    late final String v2;
    late final String legacy;
    switch (type) {
      case 'national_id':
        v2 = 'doc_national_id';
        legacy = 'img_id_rksh';
        break;
      case 'vehicle_registration':
        v2 = 'doc_vehicle_registration';
        legacy = 'img_id_car';
        break;
      case 'driver_license':
        return _driverLicenseAggregateStatus(data);
      default:
        return DriverRegistrationDocStatus.missing;
    }
    final raw = _slotStatusRaw(data, v2);
    if (raw == 'rejected') return DriverRegistrationDocStatus.rejected;
    if (raw == 'needs_reupload')
      return DriverRegistrationDocStatus.needsReupload;
    if (_hasPresentAsset(data, v2, legacy)) {
      return DriverRegistrationDocStatus.complete;
    }
    return DriverRegistrationDocStatus.missing;
  }

  static DriverRegistrationDocStatus _driverLicenseAggregateStatus(
    Map<String, dynamic> data,
  ) {
    for (final key in [
      DriverLicenseDocumentFields.front,
      DriverLicenseDocumentFields.back,
      DriverLicenseDocumentFields.legacy,
    ]) {
      final raw = _slotStatusRaw(data, key);
      if (raw == 'rejected') return DriverRegistrationDocStatus.rejected;
      if (raw == 'needs_reupload') {
        return DriverRegistrationDocStatus.needsReupload;
      }
    }
    if (DriverLicenseDocument.isCompleteForSubmit(data)) {
      return DriverRegistrationDocStatus.complete;
    }
    if (DriverLicenseDocument.isApprovedLegacyLicenseOnly(data)) {
      return DriverRegistrationDocStatus.complete;
    }
    return DriverRegistrationDocStatus.missing;
  }

  static bool profilePhotoOk(Map<String, dynamic> data) {
    final path = (data['photo_storage_path'] as String?)?.trim() ?? '';
    if (isStoragePath(path)) return true;
    final u = (data['photo_url'] as String?)?.trim() ?? '';
    return u.startsWith('https://');
  }

  /// Overall V2 required docs (excluding profile photo).
  static DriverRegistrationDocStatus overall(Map<String, dynamic> data) {
    final slots = requiredTypes.map((t) => statusForType(data, t)).toList();
    if (slots.any((s) => s == DriverRegistrationDocStatus.rejected)) {
      return DriverRegistrationDocStatus.rejected;
    }
    if (slots.any((s) => s == DriverRegistrationDocStatus.needsReupload)) {
      return DriverRegistrationDocStatus.needsReupload;
    }
    if (slots.any((s) => s == DriverRegistrationDocStatus.missing)) {
      return DriverRegistrationDocStatus.missing;
    }
    return DriverRegistrationDocStatus.complete;
  }

  static bool isComplete(Map<String, dynamic> data) =>
      overall(data) == DriverRegistrationDocStatus.complete &&
      profilePhotoOk(data);

  static List<String> missingTypes(Map<String, dynamic> data) {
    return requiredTypes
        .where((t) =>
            statusForType(data, t) != DriverRegistrationDocStatus.complete)
        .toList();
  }
}
